File: 74.cpp
#include <algorithm>
#include <deque>
#include <iostream>
#include <map>
#include <set>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;
class Solution {
public:
  bool searchMatrix(vector<vector<int>> &matrix, int target) {
    int m = matrix.size(), n = matrix[0].size();
    int l = 0, r = m * n;
    while (l < r) {
      int midPoint = (r - l) / 2 + l;
      int row = midPoint / n, col = midPoint % n;
      if (matrix[row][col] < target) {
        l = midPoint + 1;
      } else if (matrix[row][col] > target) {
        r = midPoint;
      } else {
        return true;
      }
    }
    return false;
  }
};
int main() {}

File: 74.py
class Solution:
    def map1d_to_2d(self, idx, n):
        return (idx // n, idx % n)

    def searchMatrix(self, matrix: list[list[int]], target: int) -> bool:
        m = len(matrix)
        n = len(matrix[0])
        l = 0
        r = m * n
        while l < r:
            midpoint = (r - l) // 2 + l
            row, col = self.map1d_to_2d(midpoint, n)
            if matrix[row][col] == target:
                return True
            elif matrix[row][col] > target:
                r = midpoint
            else:
                l = midpoint + 1
        return False


print(Solution().searchMatrix([[1]], 2))
